packages/drivers/odsp-driver/src/authorizedFetchWithRetry.ts

```diff
@@ -0,0 +1,163 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger, ITelemetryProperties } from "@fluidframework/common-definitions";
```

Comment:
Did not look at all the changes - code change (around the sharingLink passed in) looks good. 2 clarifications:
1. Double checking - there are no assumptions around the kind of URL that is being passed in for the sharingLink, is there? (it could be a variety of sharing links, and clients shouldnt need to build it).
2. I assume the URL is coming in from the host app directly:
---- Recipient Cases: Should alreayd have the sharing link, and it should get passed in to the driver.
---- Sender Cases: Explicitly call the service to get the sharing link and use it here if needed.
3. General question: is there some assumption around the itemId/driveId URL being passed in as the main URL? If (in the future, not this PR), we want to send in a sharing URL (for chap2 scenarios etc.), I assume that can be done?

Reply:
1.) We are not taking the sharing link from the app. Rather the driver is generating the link, only the token in passed from the app through the callback.
2.) Yes the Url(request to be resolved comes from the app directly). So sharing link can be passed through that request or if not then it can be generated in the resolver. But currently we are always generating it in the resolver.
3.) Only certain kind or urls are resolved by a given url resolver. They need to be in a particular format. What do you mean by main URL? Currently sharing link is generated by the resolver in this PR. If for some scenario, sharing link needs to be supplied by the resolver, then app can supply it in the request to be resolved(we can add the link in the format).

Reply:
2. The entire point of this feature is to avoid additional round trips for generating sharing links.

Reply:
We need to generate the sharing links in the driver for hosts who don't pass in sharing links. That said, when I first implemented this logic, I asked Andrei if there was a way to detect if the link being passed in was a sharing link already. I was told there was no way to do this. If that's the case, I don't see another way for us to avoid this call (besides putting generating a sharing link on the critical path of container/app boot which I also tried and was told was majorly undesirable for performance)

Reply:
Yes we would need a way to determine when to generate the link? @ram-ramasubramanian

Reply:
I added the headers for sharing link as we discussed.